modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/ModifyNode.java

```diff
@@ -0,0 +1,236 @@
+    @Override public void push(Row row) throws Exception {
+        assert downstream() != null;
+        assert waiting > 0;
+        assert state == State.UPDATING;
+
+        checkState();
+
+        waiting--;
+
+        switch (op) {
+            case DELETE:
+            case UPDATE:
+            case INSERT:
+                tuples.add(desc.toTuple(context(), row, op, cols));
+
+                flushTuples(false);
+
+                break;
+            default:
+                throw new UnsupportedOperationException(op.name());
+        }
+
+        if (waiting == 0)
+            source().request(waiting = MODIFY_BATCH_SIZE);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void end() throws Exception {
+        assert downstream() != null;
+        assert waiting > 0;
+
+        checkState();
+
+        waiting = -1;
+        state = State.UPDATED;
+
+        tryEnd();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void rewindInternal() {
+        throw new UnsupportedOperationException();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Downstream<Row> requestDownstream(int idx) {
+        if (idx != 0)
+            throw new IndexOutOfBoundsException();
+
+        return this;
+    }
+
+    /** */
+    private void tryEnd() throws Exception {
+        assert downstream() != null;
+
+        if (state == State.UPDATING && waiting == 0)
+            source().request(waiting = MODIFY_BATCH_SIZE);
+
+        if (state == State.UPDATED && requested > 0) {
+            flushTuples(true);
+
+            state = State.END;
+
+            inLoop = true;
+            try {
+                requested--;
+                downstream().push(context().rowHandler().factory(long.class).create(updatedRows));
+            }
+            finally {
+                inLoop = false;
+            }
+        }
+
+        if (state == State.END && requested > 0) {
+            requested = 0;
+            downstream().end();
+        }
+    }
+
+    /** */
+    private void flushTuples(boolean force) {
+        if (nullOrEmpty(tuples) || !force && tuples.size() < MODIFY_BATCH_SIZE)
+            return;
+
+        List<Tuple> tuples = this.tuples;
+        this.tuples = new ArrayList<>(MODIFY_BATCH_SIZE);
+
+        switch (op) {
+            case INSERT:
+                Collection<Tuple> duplicates = tableView.insertAll(tuples);
+
+                if (!duplicates.isEmpty()) {
+                    IgniteTypeFactory typeFactory = context().getTypeFactory();
+                    RowHandler.RowFactory<Row> rowFactory = context().rowHandler().factory(
+                        context().getTypeFactory(),
+                        desc.insertRowType(typeFactory)
+                    );
+
+                    throw new IgniteInternalException("Failed to INSERT some keys because they are already in cache. " +
+                        "[tuples=" + duplicates.stream().map(tup -> desc.toRow(context(), tup, rowFactory, null))
```

Comment:
please put each stream call on separate line and align ones

Reply:
done